File: api/app/domain/suspects.py
"""Which ingredients kept showing up before this week's flares."""

from dataclasses import dataclass, field
from datetime import datetime

from app.domain.flares import Flare, cluster_flares, hours_between, meals_before
from app.domain.foods import meal_items, normalize_item
from app.domain.rounding import round1, round2
from app.domain.rows import CorrelationRow, MealRow, SymptomRow
from app.domain.time import local_date, week_dates

ALL_SYMPTOMS = "All Symptoms"
TOP_SUSPECTS = 5


@dataclass
class SuspectPair:
    meal_id: int
    meal_name: str
    meal_at: datetime
    flare_at: datetime
    symptoms: list[str]
    onset_hours: float


@dataclass
class Suspect:
    name: str
    flares_with_ingredient: int
    flares_total: int
    share: float
    avg_onset_hours: float | None
    times_logged_this_week: int
    exposures_all_time: int
    confidence: int
    on_watchlist: bool
    recent_pairs: list[SuspectPair]


@dataclass
class TimingWindow:
    flares: int = 0
    top_ingredients: list[str] = field(default_factory=list)


@dataclass
class SuspectsDigest:
    week_start: str
    week_end: str
    window_hours: int
    flares: int
    meals_evaluated: int
    lead_suspect: str | None
    symptom_filters: list[dict]
    ingredients: list[Suspect]
    timing_windows: dict[str, TimingWindow]


@dataclass
class _Entry:
    display: str
    hits: list[tuple[Flare, MealRow, float]] = field(default_factory=list)
    flares: set = field(default_factory=set)


def compute_suspects(
    meals: list[MealRow],
    symptoms: list[SymptomRow],
    correlations: list[CorrelationRow],
    watchlist: list[str],
    week_start: str,
    tz: str,
    window_hours: int,
    symptom_filter: str | None = None,
) -> SuspectsDigest:
    dates = set(week_dates(week_start))
    week_symptoms = [s for s in symptoms if local_date(s.timestamp, tz) in dates]
    all_flares = cluster_flares(week_symptoms)

    symptom_filters = [{"name": ALL_SYMPTOMS, "count": len(all_flares)}]
    for name in sorted({s.name for s in week_symptoms}):
        symptom_filters.append(
            {"name": name, "count": sum(1 for f in all_flares if name in f.names)}
        )

    flares = (
        [f for f in all_flares if symptom_filter in f.names] if symptom_filter else all_flares
    )
    watched = {normalize_item(item) for item in watchlist}

    entries: dict[str, _Entry] = {}
    evaluated: set[int] = set()
    for flare in flares:
        for meal in meals_before(meals, flare.start, window_hours):
            evaluated.add(meal.id)
            onset = hours_between(meal.timestamp, flare.start)
            for item in meal_items(meal):
                if item.dimension != "ingredient":
                    continue
                entry = entries.setdefault(item.key, _Entry(display=item.display))
                # Only the most recent meal of the window sets the onset
                if flare not in entry.flares:
                    entry.hits.append((flare, meal, onset))
                entry.flares.add(flare)

    week_meals = [m for m in meals if local_date(m.timestamp, tz) in dates]

    def count_with(pool: list[MealRow], key: str) -> int:
        return sum(
            1
            for meal in pool
            if any(i.dimension == "ingredient" and i.key == key for i in meal_items(meal))
        )

    def confidence_for(key: str) -> int:
        matching = [
            row.confidence
            for row in correlations
            if row.dimension == "ingredient"
            and normalize_item(row.food_name) == key
            and (not symptom_filter or row.symptom_name == symptom_filter)
        ]
        return max(matching, default=0)

    suspects = [
        Suspect(
            name=entry.display,
            flares_with_ingredient=len(entry.flares),
            flares_total=len(flares),
            share=round2(len(entry.flares) / len(flares)) if flares else 0,
            avg_onset_hours=round1(sum(h for _, _, h in entry.hits) / len(entry.hits))
            if entry.hits
            else None,
            times_logged_this_week=count_with(week_meals, key),
            exposures_all_time=count_with(meals, key),
            confidence=confidence_for(key),
            on_watchlist=key in watched,
            recent_pairs=[
                SuspectPair(
                    meal_id=meal.id,
                    meal_name=meal.name,
                    meal_at=meal.timestamp,
                    flare_at=flare.start,
                    symptoms=flare.names,
                    onset_hours=round1(onset),
                )
                for flare, meal, onset in sorted(
                    entry.hits, key=lambda hit: hit[0].start, reverse=True
                )[:3]
            ],
        )
        for key, entry in entries.items()
    ]
    suspects.sort(key=lambda s: (-s.share, -s.confidence, s.name.casefold(), s.name))
    suspects = suspects[:TOP_SUSPECTS]

    # Each flare is attributed to its most recent prior meal
    buckets: dict[str, tuple[TimingWindow, dict[str, int]]] = {
        "0to4h": (TimingWindow(), {}),
        "4to12h": (TimingWindow(), {}),
        "12to24h": (TimingWindow(), {}),
    }
    for flare in flares:
        prior = meals_before(meals, flare.start, window_hours)
        if not prior:
            continue
        onset = hours_between(prior[0].timestamp, flare.start)
        name = "0to4h" if onset <= 4 else "4to12h" if onset <= 12 else "12to24h"
        window, counts = buckets[name]
        window.flares += 1
        for item in meal_items(prior[0]):
            if item.dimension == "ingredient":
                counts[item.display] = counts.get(item.display, 0) + 1
    for window, counts in buckets.values():
        window.top_ingredients = [
            name for name, _ in sorted(counts.items(), key=lambda kv: -kv[1])[:3]
        ]

    return SuspectsDigest(
        week_start=week_start,
        week_end=week_dates(week_start)[6],
        window_hours=window_hours,
        flares=len(flares),
        meals_evaluated=len(evaluated),
        lead_suspect=suspects[0].name if suspects else None,
        symptom_filters=symptom_filters,
        ingredients=suspects,
        timing_windows={name: window for name, (window, _) in buckets.items()},
    )

File: app/server/analytics/suspects.ts
import { Correlation, Meal, Symptom } from "@shared/schema";
import { localDate } from "./time";
import { clusterFlares, Flare, hoursBetween, mealsBefore } from "./flares";
import { mealItems, normalizeItem } from "./foods";
import { weekDates } from "./digest";

export type SuspectPair = {
  mealId: number;
  mealName: string;
  mealAt: Date;
  flareAt: Date;
  symptoms: string[];
  onsetHours: number;
};

export type Suspect = {
  name: string;
  flaresWithIngredient: number;
  flaresTotal: number;
  share: number;
  avgOnsetHours: number | null;
  timesLoggedThisWeek: number;
  exposuresAllTime: number;
  confidence: number;
  onWatchlist: boolean;
  recentPairs: SuspectPair[];
};

export type TimingWindow = { flares: number; topIngredients: string[] };

export type SuspectsDigest = {
  weekStart: string;
  weekEnd: string;
  windowHours: number;
  flares: number;
  mealsEvaluated: number;
  leadSuspect: string | null;
  symptomFilters: { name: string; count: number }[];
  ingredients: Suspect[];
  timingWindows: { "0to4h": TimingWindow; "4to12h": TimingWindow; "12to24h": TimingWindow };
};

function round1(n: number): number {
  return Math.round(n * 10) / 10;
}

/**
 * Which ingredients showed up in the window before this week's flares.
 * `meals`/`symptoms` are full history; `symptomFilter` narrows the flares.
 */
export function computeSuspects(
  meals: Meal[],
  symptoms: Symptom[],
  correlations: Correlation[],
  watchlist: string[],
  weekStart: string,
  tz: string,
  windowHours: number,
  symptomFilter?: string | null,
): SuspectsDigest {
  const dates = weekDates(weekStart);
  const inWeek = new Set(dates);
  const weekSymptoms = symptoms.filter((s) => inWeek.has(localDate(s.timestamp, tz)));
  const allFlares = clusterFlares(weekSymptoms);

  const symptomFilters = [{ name: "All Symptoms", count: allFlares.length }];
  for (const name of [...new Set(weekSymptoms.map((s) => s.name))].sort()) {
    symptomFilters.push({ name, count: allFlares.filter((f) => f.names.includes(name)).length });
  }

  const flares = symptomFilter ? allFlares.filter((f) => f.names.includes(symptomFilter)) : allFlares;
  const watch = new Set(watchlist.map(normalizeItem));

  // Ingredient exposures before each flare
  type Hit = { flare: Flare; meal: Meal; onset: number };
  const hitsByItem = new Map<string, { display: string; hits: Hit[]; flares: Set<Flare> }>();
  const evaluated = new Set<number>();
  for (const flare of flares) {
    for (const meal of mealsBefore(meals, flare.start, windowHours)) {
      evaluated.add(meal.id);
      const onset = hoursBetween(meal.timestamp, flare.start);
      for (const item of mealItems(meal).filter((i) => i.dimension === "ingredient")) {
        const entry = hitsByItem.get(item.key) ?? { display: item.display, hits: [], flares: new Set() };
        // Only the most recent exposure per flare counts for onset
        if (!entry.flares.has(flare)) entry.hits.push({ flare, meal, onset });
        entry.flares.add(flare);
        hitsByItem.set(item.key, entry);
      }
    }
  }

  const weekMeals = meals.filter((m) => inWeek.has(localDate(m.timestamp, tz)));
  const countItem = (list: Meal[], key: string) => list.filter((m) => mealItems(m).some((i) => i.dimension === "ingredient" && i.key === key)).length;
  const confidenceFor = (key: string) => Math.max(0, ...correlations
    .filter((c) => c.dimension === "ingredient" && normalizeItem(c.foodName) === key && (!symptomFilter || c.symptomName === symptomFilter))
    .map((c) => c.confidence));

  const ingredients: Suspect[] = [...hitsByItem.entries()]
    .map(([key, entry]) => ({
      name: entry.display,
      flaresWithIngredient: entry.flares.size,
      flaresTotal: flares.length,
      share: flares.length ? Math.round(entry.flares.size / flares.length * 100) / 100 : 0,
      avgOnsetHours: entry.hits.length ? round1(entry.hits.reduce((s, h) => s + h.onset, 0) / entry.hits.length) : null,
      timesLoggedThisWeek: countItem(weekMeals, key),
      exposuresAllTime: countItem(meals, key),
      confidence: confidenceFor(key),
      onWatchlist: watch.has(key),
      recentPairs: entry.hits
        .sort((a, b) => b.flare.start.getTime() - a.flare.start.getTime())
        .slice(0, 3)
        .map((h) => ({ mealId: h.meal.id, mealName: h.meal.name, mealAt: h.meal.timestamp, flareAt: h.flare.start, symptoms: h.flare.names, onsetHours: round1(h.onset) })),
    }))
    .sort((a, b) => b.share - a.share || b.confidence - a.confidence || a.name.localeCompare(b.name))
    .slice(0, 5);

  // Timing windows by the most recent meal before each flare
  const windows = { "0to4h": { flares: 0, items: new Map<string, number>() }, "4to12h": { flares: 0, items: new Map<string, number>() }, "12to24h": { flares: 0, items: new Map<string, number>() } };
  for (const flare of flares) {
    const [latest] = mealsBefore(meals, flare.start, windowHours);
    if (!latest) continue;
    const onset = hoursBetween(latest.timestamp, flare.start);
    const bucket = onset <= 4 ? windows["0to4h"] : onset <= 12 ? windows["4to12h"] : windows["12to24h"];
    bucket.flares++;
    for (const item of mealItems(latest).filter((i) => i.dimension === "ingredient")) {
      bucket.items.set(item.display, (bucket.items.get(item.display) ?? 0) + 1);
    }
  }
  const finish = (w: { flares: number; items: Map<string, number> }): TimingWindow => ({
    flares: w.flares,
    topIngredients: [...w.items.entries()].sort((a, b) => b[1] - a[1]).slice(0, 3).map(([name]) => name),
  });

  return {
    weekStart,
    weekEnd: dates[6],
    windowHours,
    flares: flares.length,
    mealsEvaluated: evaluated.size,
    leadSuspect: ingredients[0]?.name ?? null,
    symptomFilters,
    ingredients,
    timingWindows: { "0to4h": finish(windows["0to4h"]), "4to12h": finish(windows["4to12h"]), "12to24h": finish(windows["12to24h"]) },
  };
}
